src/shared/mkfs-util.c

```diff
@@ -147,15 +148,28 @@ static int do_mcopy(const char *node, const char *root) {
         /* mcopy copies the top level directory instead of everything in it so we have to pass all
          * the subdirectories to mcopy instead to end up with the correct directory structure. */
 
-        rootdir = opendir(root);
-        if (!rootdir)
-                return log_error_errno(errno, "Failed to open directory '%s'", root);
+        rfd = open(root, O_RDONLY|O_DIRECTORY|O_CLOEXEC);
+        if (rfd < 0)
+                return log_error_errno(errno, "Failed to open directory '%s': %m", root);
 
-        FOREACH_DIRENT(de, rootdir, return -errno) {
-                char *p = path_join(root, de->d_name);
+        r = readdir_all(rfd, RECURSE_DIR_SORT, &de);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read '%s' contents: %m", root);
+
+        for (size_t i = 0; i < de->n_entries; i++) {
+                char *p = path_join(root, de->entries[i]->d_name);
                 if (!p)
                         return log_oom();
 
+                r = dirent_ensure_type(rfd, de->entries[i]);
```

Comment:
How about to add a new flag to `readdir_all()`, say `RECURSE_DIR_ENSURE_TYPE`?

Reply:
This actually already exists

Reply:
Indeed, but `readdir_all()` seems to not use the flag...

Reply:
Pushed a new update that makes readdir_all() handle RECURSE_DIR_ENSURE_TYPE